### ISSUE TITLE
AY 19/20 Sem2 Qn4 : Exceptions Ordering.

### ISSUE THREAD

**Issue by Herrekt** [nus-cs2030/2021-s2#389]
# Summary
I know that we should focus on the more specific exceptions to the least specific ones, but if they branch off from the same parent exception, can we clear one whole child exception first before starting on the next?

## Description 
For example, if we were to catch NoPermissionException, then catch WriteException, before catching ElementNotFoundException and Read Exception, will it be problematic/wrong in this case?

## Screenshots (if any): 
![Capture](https://user-images.githubusercontent.com/80625519/116215974-ba659400-a77a-11eb-838a-06a9ec81b18e.PNG)
![Capture1](https://user-images.githubusercontent.com/80625519/116217462-3ca28800-a77c-11eb-9606-b2d9165fa81e.PNG)

**Comment by brendancjz** [comment]
Hi, I believe either way is fine. The more common approach will be to the finish catching the child exception of a branch. As long as you are always catching the next most specific exception, you're good.

**Comment by raihanyusri** [comment]
Either way is fine. If the exceptions have a parent-child relationship, then the catch blocks must be sorted by the most specific exceptions first, followed by the most general ones within the group itself

**Comment by gableejh** [comment]
both works or you can use `|` to catch both, refer to slide 10 of lecture 6

**Comment by liongthryan** [comment]
I think your method works fine! If you're writing the code in an IDE, most IDEs flag unreachable code blocks (ie. you tried to catch a more general exception before catching the specific exception). Since prof said we can write our answers for finals using any IDEs (iirc), it might not be a bad strategy to just rely on those!

**Comment by deniselam10** [comment]
I think its fine! As long as the later exception is not more specific than the one caught previously, your code should work.

**Comment by yowyuxuan** [comment]
I did it in a similar way as well! so long it is bottom up! :)

**Comment by MatthiasTanWeiSheng** [comment]
i think both is fine. As long as you are catching the more specific exceptions first.

**Comment by yx0555** [comment]
I did it the same way as you too, should work since the more specific exceptions are called first!

**Comment by criyu95** [comment]
Both ways are fine, as long as the more specific exceptions are caught first

**Comment by amaslua** [comment]
It should work, i did it the same way as u! as long as more specific to less specific

**Comment by jeremycte** [comment]
Hi @Herrekt , you can view more in-depth based on my explaination here #307 but basically either way is fine. The primary goal is to catch the more specific exceptions/ child exceptions before catching the general exceptions/parent exceptions last. 😃

**Comment by linminsaw** [comment]
In general, as long as you catch the exceptions from the most specific to the most general, it will work well :)

**Comment by YuanZhengyz** [comment]
Yup as long as you don't accidentally check for a more general exception first it will be fine.

**Comment by andreathr** [comment]
I think as long as you catch the most specific exception which is the NoPermissionException its fine!

**Comment by Jjjetplane** [comment]
As long as you are catching them from the most specific to the most general it will work!

**Comment by gwajoon** [comment]
It will work as long as you catch them in order of specificity, starting from the most specific one!

**Comment by dannytayjy** [comment]
```
try { 
    accessDatabase();
} catch (NoPermissionException e1) {
    System.out.print("NoPermissionException");
} catch (ElementNotFoundException e2) { 
    System.out.print("ElementNotFoundException");
} catch (WriteException e3) {
    System.out.print("WriteException");
} catch (ReadException e4) {
    System.out.print("ReadException");
} catch (TableException e5) {
    System.out.print("TableException");
} catch (DatabaseException e6) {
    System.out.print("DatabaseException");
}
```

**Comment by NIL-99** [comment]
rule of thumb is to just catch the more specific errors than the more general errors. Arrangement not be a problem unless you catch a more general error before a more specific error instead

**Comment by W-ShiHan** [comment]
I think that both your answers work.

Personally I draw out my workings in a tree, 
then you can either 
1. Check exceptions in the order of their distance away from the root (furthest away first)
2. Start from any "leaf" of the tree and check for exceptions backwards, and repeat until all exceptions have been checked.
(sorry for my poor handwriting)
![tree](https://user-images.githubusercontent.com/77675146/116368745-3fb07d80-a83b-11eb-911b-86c8d4135dc9.PNG)

**Comment by JWulaXia** [comment]
idea is go from the most specific exception, if two exception are at the same level, the the order does not matter. this can ensure all exceptions are caught and are caught in the most specific one
